Get-ChildItem of (Get-Volume) Paths does not work as expected

### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

During the creation of a module, I run back and forth between 5.1 and pwsh (currently 7.1.3) to make sure the module will work in both versions.

While running in PWSH-7.1.3:

$ISO = 'C:\WindowsImages\Win10_21H1_English_x64.iso'
$Drive = Mount-DiskImage $ISO -StorageType ISO -NoDriveLetter -PassThru
gci -LiteralPath (Get-Volume | ?{($_.DriveType -match 'CD') -and ($_.Size -match $Drive.Size)}).Path -fo -re

I get back a live list of all the files found in 'C:\'


### Expected behavior

```console
In powershell 5.1, I get back all the files found in:

Directory: \\?\Volume{1a1cbe60-f411-11eb-aec7-001a7dda7113}
```


### Actual behavior

```console
In PWSH 7.1.3, I get back all the files found in:

Directory: C:\
```


### Error details

```console
No Errors, I just dont get the correct results path in PWSH
```


### Environment data

```powershell
Name                           Value                                                                                                  
----                           -----                                                                                                  
PSVersion                      5.1.19041.1023                                                                                         
PSEdition                      Desktop                                                                                                
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                
BuildVersion                   10.0.19041.1023                                                                                        
CLRVersion                     4.0.30319.42000                                                                                        
WSManStackVersion              3.0                                                                                                    
PSRemotingProtocolVersion      2.3                                                                                                    
SerializationVersion           1.1.0.1                                                                                                


&



Name                           Value
----                           -----
PSVersion                      7.1.3
PSEdition                      Core
GitCommitId                    7.1.3
OS                             Microsoft Windows 10.0.19043
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0ΓÇª}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```


### Visuals

Powershell 5.1

![image](https://user-images.githubusercontent.com/71091773/128242532-a1e4568a-c4d8-4042-b977-e9e9f0a93586.png)

Powershell 7.1.3

![image](https://user-images.githubusercontent.com/71091773/128242687-7078d35c-31fc-42e8-bdfc-41a9cc5dfeb5.png)

I suspect this may be related to an issue I'm seeing in PowerShell 7.2.x, I don't recall this in 7.0.x.  `Get-Volume` ocassionally doesn't return any results.

```powershell
C:\Users\Fots> Get-Volume

DriveLetter FriendlyName       FileSystemType DriveType HealthStatus OperationalStatus SizeRemaining      Size
----------- ------------       -------------- --------- ------------ ----------------- -------------      ----
D           Overflow           NTFS           Fixed     Healthy      OK                    199.57 GB 687.37 GB
E           Kontakt Libraries  NTFS           Fixed     Healthy      OK                    262.43 GB   1.29 TB
            Work Windows 10 NTFS           Fixed     Healthy      OK                     46.76 GB 244.12 GB
H           Fotsies Backup 2   NTFS           Fixed     Healthy      OK                      2.78 TB   3.64 TB
I           Fotsies Backup 1   NTFS           Fixed     Healthy      OK                      2.78 TB   3.64 TB
F           Sample Packs       NTFS           Fixed     Healthy      OK                     18.71 GB  56.37 GB
C           Windows 11         NTFS           Fixed     Healthy      OK                    104.52 GB 488.17 GB
G           Backup             NTFS           Fixed     Healthy      OK                    753.89 GB   1.82 TB
```

As you can see, the first call to `Get-Volume` returned nothing.  This is causing my scripts to fail on the first run.

@SteveL-MSFT Can you confirm if this is the same issue or does my problem look different?  Just went on the hunt for similar issues before opening a fresh one.

Huge thanks
Fotis

--- a/src/System.Management.Automation/engine/Utils.cs
+++ b/src/System.Management.Automation/engine/Utils.cs
@@ -1292,7 +1292,7 @@ namespace System.Management.Automation
             }
 
             // handle special cases like \\wsl$\ubuntu and \\?\ which aren't a UNC path, but we can say it is so the filesystemprovider can use it
-            if (!networkOnly && (path.StartsWith(WslRootPath, StringComparison.OrdinalIgnoreCase) || path.StartsWith("\\\\?\\")))
+            if (!networkOnly && (path.StartsWith(WslRootPath, StringComparison.OrdinalIgnoreCase) || path.StartsWith("\\\\?\\") || path.StartsWith("\\\\.\\")))
             {
                 return true;
             }

--- a/src/System.Management.Automation/engine/Utils.cs
+++ b/src/System.Management.Automation/engine/Utils.cs
@@ -1291,7 +1291,7 @@ namespace System.Management.Automation
                 return false;
             }
 
-            // handle special cases like \\wsl$\ubuntu and \\?\ which aren't a UNC path, but we can say it is so the filesystemprovider can use it
+            // handle special cases like '\\wsl$\ubuntu', '\\?\', and '\\.\pipe\' which aren't a UNC path, but we can say it is so the filesystemprovider can use it
             if (!networkOnly && (path.StartsWith(WslRootPath, StringComparison.OrdinalIgnoreCase) || path.StartsWith("\\\\?\\") || path.StartsWith("\\\\.\\")))
             {
                 return true;

--- a/src/System.Management.Automation/engine/Utils.cs
+++ b/src/System.Management.Automation/engine/Utils.cs
@@ -1291,8 +1291,8 @@ namespace System.Management.Automation
                 return false;
             }
 
-            // handle special cases like \\wsl$\ubuntu which isn't a UNC path, but we can say it is so the filesystemprovider can use it
-            if (path.StartsWith(WslRootPath, StringComparison.OrdinalIgnoreCase))
+            // handle special cases like \\wsl$\ubuntu and \\?\ which aren't a UNC path, but we can say it is so the filesystemprovider can use it
+            if (path.StartsWith(WslRootPath, StringComparison.OrdinalIgnoreCase) || path.StartsWith("\\\\?\\"))
             {
                 return true;
             }

--- a/src/System.Management.Automation/engine/Utils.cs
+++ b/src/System.Management.Automation/engine/Utils.cs
@@ -1281,7 +1281,7 @@ namespace System.Management.Automation
             return false;
         }
 
-        internal static bool PathIsUnc(string path)
+        internal static bool PathIsUnc(string path, bool networkOnly = false)
         {
 #if UNIX
             return false;
@@ -1292,7 +1292,7 @@ namespace System.Management.Automation
             }
 
             // handle special cases like \\wsl$\ubuntu and \\?\ which aren't a UNC path, but we can say it is so the filesystemprovider can use it
-            if (path.StartsWith(WslRootPath, StringComparison.OrdinalIgnoreCase) || path.StartsWith("\\\\?\\"))
+            if (!networkOnly && (path.StartsWith(WslRootPath, StringComparison.OrdinalIgnoreCase) || path.StartsWith("\\\\?\\")))
             {
                 return true;
             }

--- a/src/System.Management.Automation/namespaces/FileSystemProvider.cs
+++ b/src/System.Management.Automation/namespaces/FileSystemProvider.cs
@@ -124,7 +124,7 @@ namespace Microsoft.PowerShell.Commands
             {
                 string exactPath = string.Empty;
                 int itemsToSkip = 0;
-                if (Utils.PathIsUnc(path))
+                if (Utils.PathIsUnc(path) || path.StartsWith("\\\\?\\"))
                 {
                     // With the Split method, a UNC path like \\server\share, we need to skip
                     // trying to enumerate the server and share, so skip the first two empty
@@ -132,20 +132,21 @@ namespace Microsoft.PowerShell.Commands
                     itemsToSkip = 4;
                 }
 
-                foreach (string item in path.Split(StringLiterals.DefaultPathSeparator))
+                var items = path.Split(StringLiterals.DefaultPathSeparator);
+                for (int i = 0; i < items.Length; i++)
                 {
                     if (itemsToSkip-- > 0)
                     {
                         // This handles the UNC server and share and 8.3 short path syntax
-                        exactPath += item + StringLiterals.DefaultPathSeparator;
+                        exactPath += items[i] + StringLiterals.DefaultPathSeparator;
                         continue;
                     }
                     else if (string.IsNullOrEmpty(exactPath))
                     {
                         // This handles the drive letter or / root path start
-                        exactPath = item + StringLiterals.DefaultPathSeparator;
+                        exactPath = items[i] + StringLiterals.DefaultPathSeparator;
                     }
-                    else if (string.IsNullOrEmpty(item))
+                    else if (string.IsNullOrEmpty(items[i]) && i == items.Length - 1)
                     {
                         // This handles the trailing slash case
                         if (!exactPath.EndsWith(StringLiterals.DefaultPathSeparator))
@@ -155,17 +156,17 @@ namespace Microsoft.PowerShell.Commands
 
                         break;
                     }
-                    else if (item.Contains('~'))
+                    else if (items[i].Contains('~'))
                     {
                         // This handles short path names
-                        exactPath += StringLiterals.DefaultPathSeparator + item;
+                        exactPath += StringLiterals.DefaultPathSeparator + items[i];
                     }
                     else
                     {
                         // Use GetFileSystemEntries to get the correct casing of this element
                         try
                         {
-                            var entries = Directory.GetFileSystemEntries(exactPath, item);
+                            var entries = Directory.GetFileSystemEntries(exactPath, items[i]);
                             if (entries.Length > 0)
                             {
                                 exactPath = entries[0];

--- a/src/System.Management.Automation/namespaces/FileSystemProvider.cs
+++ b/src/System.Management.Automation/namespaces/FileSystemProvider.cs
@@ -124,7 +124,7 @@ namespace Microsoft.PowerShell.Commands
             {
                 string exactPath = string.Empty;
                 int itemsToSkip = 0;
-                if (Utils.PathIsUnc(path) || path.StartsWith("\\\\?\\"))
+                if (Utils.PathIsUnc(path))
                 {
                     // With the Split method, a UNC path like \\server\share, we need to skip
                     // trying to enumerate the server and share, so skip the first two empty

--- a/src/System.Management.Automation/namespaces/FileSystemProvider.cs
+++ b/src/System.Management.Automation/namespaces/FileSystemProvider.cs
@@ -7258,7 +7258,7 @@ namespace Microsoft.PowerShell.Commands
                     return false;
                 }
 
-                if (Utils.PathIsUnc(path))
+                if (Utils.PathIsUnc(path, networkOnly : true))
                 {
                     return true;
                 }

--- a/test/powershell/Modules/Microsoft.PowerShell.Management/Get-ChildItem.Tests.ps1
+++ b/test/powershell/Modules/Microsoft.PowerShell.Management/Get-ChildItem.Tests.ps1
@@ -235,6 +235,11 @@ Describe "Get-ChildItem" -Tags "CI" {
             $items[0].Parent | Should -BeExactly "${volume}Windows"
             $items.Count | Should -Be (Get-ChildItem $env:SystemRoot).Count
         }
+
+        It 'Works with Windows pipes' -Skip:(!$IsWindows) {
+            $out = pwsh -noprofile -custompipename myTestPipe { Get-ChildItem \\.\pipe\myTestPipe }
+            $out.Name | Should -BeExactly 'myTestPipe'
+        }
     }
 
     Context 'Env: Provider' {

--- a/test/powershell/Modules/Microsoft.PowerShell.Management/Get-ChildItem.Tests.ps1
+++ b/test/powershell/Modules/Microsoft.PowerShell.Management/Get-ChildItem.Tests.ps1
@@ -228,6 +228,13 @@ Describe "Get-ChildItem" -Tags "CI" {
             $barFiles = Get-ChildItem -Path testdrive:/foo/bar -Recurse
             $barFiles.Count | Should -Be 2
         }
+
+        It 'Works with Windows volume paths' -Skip:(!$IsWindows) {
+            $volume = (Get-Volume -DriveLetter $env:SystemDrive[0]).Path
+            $items = Get-ChildItem -LiteralPath "${volume}Windows"
+            $items[0].Parent | Should -BeExactly "${volume}Windows"
+            $items.Count | Should -Be (Get-ChildItem $env:SystemRoot).Count
+        }
     }
 
     Context 'Env: Provider' {

--- a/test/powershell/Modules/Microsoft.PowerShell.Management/Get-ChildItem.Tests.ps1
+++ b/test/powershell/Modules/Microsoft.PowerShell.Management/Get-ChildItem.Tests.ps1
@@ -233,7 +233,7 @@ Describe "Get-ChildItem" -Tags "CI" {
             $volume = (Get-Volume -DriveLetter $env:SystemDrive[0]).Path
             $items = Get-ChildItem -LiteralPath "${volume}Windows"
             $items[0].Parent | Should -BeExactly "${volume}Windows"
-            $items.Count | Should -Be (Get-ChildItem $env:SystemRoot).Count
+            $items | Should -HaveCount (Get-ChildItem $env:SystemRoot).Count
         }
 
         It 'Works with Windows pipes' -Skip:(!$IsWindows) {

